<a href="https://colab.research.google.com/github/VindiSovia/Machine-Learning-and-Deep-Learning/blob/main/04_01_Fire_Vindi_Sovia_Anggita.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [ ]:
import pandas as pd
import numpy as np

## Input Dataset

In [ ]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [ ]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [ ]:
df.head()

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

In [ ]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography',	'Gender',	'Age', 'Tenure',	'Balance',	'NumOfProducts'	,'HasCrCard'	,'IsActiveMember'	,'EstimatedSalary',	'Exited']].copy()
df.head()

Geography  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0    France  Female   42       2       0.00              1          1   
1     Spain  Female   41       1   83807.86              1          0   
2    France  Female   42       8  159660.80              3          1   
3    France  Female   39       1       0.00              2          0   
4     Spain  Female   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  
0               1        101348.88       1  
1               1        112542.58       0  
2               0        113931.57       1  
3               0         93826.63       0  
4               1         79084.10       0

## Preprocessing

In [ ]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df)
df.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Exited  Geography_France  Geography_Germany  \
0        101348.88       1                 1                  0   
1        112542.58       0                 0                  0   
2        113931.57       1                 1                  0   
3         93826.63       0                 1                  0   
4         79084.10       0                 0                  0   

   Geography_Spain  Gender_Female  Gender_Male  
0                0              1            0  
1                1              1            0  
2                0              1            0  
3                0              1            0  
4                1              1            0

In [ ]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df.drop('Exited', axis=1)
y = df['Exited']

In [ ]:
#melihat isi X
X.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0        101348.88                 1                  0                0   
1        112542.58                 0                  0                1   
2        113931.57                 1                  0                0   
3         93826.63                 1                  0                0   
4         79084.10                 0                  0                1   

   Gender_Female  Gender_Male  
0              1            0  
1              1            0  
2              1            0  
3              1            0  
4              1            0

In [ ]:
#melihat isi Y
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [ ]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [ ]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()
X_transform.head()

Age  Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0  0.324324     0.2  0.000000       0.000000        1.0             1.0   
1  0.310811     0.1  0.334031       0.000000        0.0             1.0   
2  0.324324     0.8  0.636357       0.666667        1.0             0.0   
3  0.283784     0.1  0.000000       0.333333        0.0             0.0   
4  0.337838     0.2  0.500246       0.000000        1.0             1.0   

   EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0         0.506735               1.0                0.0              0.0   
1         0.562709               0.0                0.0              1.0   
2         0.569654               1.0                0.0              0.0   
3         0.469120               1.0                0.0              0.0   
4         0.395400               0.0                0.0              1.0   

   Gender_Female  Gender_Male  
0            1.0          0.0  
1            1.0          0.0  
2            1.0          0.0  
3            1.0          0.0  
4            1.0          0.0

## Train-Test Split

In [ ]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

## Adaboost Model

Metode ini bekerja dengan menggabungkan beberapa model yang lemah menjadi satu model yang lebih kuat. Ide dasar dibalik AdaBoost adalah untuk memberikan bobot yang berbeda pada setiap data pelatihan dan memfokuskan pelatihan pada data yang lebih sulit diprediksi oleh model sebelumnya.

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.ensemble import AdaBoostClassifier
model1 = AdaBoostClassifier()
params = {'learning_rate': [1.0, 3.0, 5.0, 7.0],
          'n_estimators': [1, 2, 3, 4]}


#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model1,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'learning_rate': 1.0, 'n_estimators': 4}

In [ ]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.97      0.91      1983
           1       0.78      0.39      0.52       517

    accuracy                           0.85      2500
   macro avg       0.82      0.68      0.72      2500
weighted avg       0.84      0.85      0.83      2500


[[1926   57]
 [ 313  204]]

0.852


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

## Support Vector Machine (SVM)

Algoritma pembelajaran mesin yang digunakan untuk tugas klasifikasi dan regresi. SVM awalnya dikembangkan untuk masalah klasifikasi dan merupakan salah satu algoritma klasifikasi yang paling kuat. Ini adalah pendekatan yang sangat berguna untuk menyelesaikan masalah klasifikasi baik dalam kasus klasifikasi biner (dua kelas) maupun dalam masalah klasifikasi kelas jamak (lebih dari dua kelas).

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.svm import SVC
model2 = SVC()
params = { "tol": [0.1,0.01,0.001],
          'C':[0.5,1.0,1.5,2.0]}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model2,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'C': 2.0, 'tol': 0.01}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.84      0.99      0.91      1983
           1       0.86      0.29      0.43       517

    accuracy                           0.84      2500
   macro avg       0.85      0.64      0.67      2500
weighted avg       0.85      0.84      0.81      2500


[[1958   25]
 [ 367  150]]

0.8432


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

## Random Forest

Algoritma pembelajaran Mechine Learning yang digunakan untuk tugas klasifikasi dan regresi. Random forest adalah jenis algoritma yang terdiri dari beberapa pohon keputusan yang bekerja sama untuk meningkatkan kinerja prediksi.

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.ensemble import RandomForestClassifier
model3 = RandomForestClassifier()
params = {'max_depth': [8, 12, 14],
          'random_state': [24, 42],
          'n_estimators': [25, 50, 100],
          'criterion': ['gini', 'entropy']}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model3,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'criterion': 'entropy',
 'max_depth': 12,
 'n_estimators': 100,
 'random_state': 24}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.79      0.46      0.58       517

    accuracy                           0.86      2500
   macro avg       0.83      0.71      0.75      2500
weighted avg       0.86      0.86      0.85      2500


[[1919   64]
 [ 280  237]]

0.8624


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

Berdasarkan 3 model yang dijalankan, Random forest memiliki score yang lebih baik dibanding Adaboost dan SVM. Random forest mendapatkan score 86%, SVM mendapatkan score 84% sedangkan Adaboost mendapatkan score 85%. Random forest mendapatkan score paling tinggi karena memastikan bahwa seluruh fitur
berkontribusi untuk menentukan hasil dari prediksi yang dilakukan. Semakin
banyak kita menentukan jumlah pohon, maka semakin tinggi akurasi yang bisa kita
peroleh dari hasil prediksi.